README.md
# Imgix Dart Client

<!-- ix-docs-ignore -->
![imgix logo](https://assets.imgix.net/sdk-imgix-logo.svg)

A Dart client library for generating URLs with imgix.

[![pub](https://badgen.net/pub/v/imgix_client)](https://pub.dev/packages/imgix_client)
![Build](https://github.com/amondnet/imgix.dart/workflows/Build/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/amondnet/imgix.dart/branch/master/graph/badge.svg)](https://codecov.io/gh/amondnet/imgix.dart)
[![License](https://img.shields.io/github/license/amondnet/imgix.dart)](https://github.com/amondnet/imgix.dart/blob/master/LICENSE)
[![HitCount](http://hits.dwyl.com/amondnet/imgix_client.svg)](http://hits.dwyl.com/amondnet/imgix_client)

---
- [Install](#install)
- [Basic Usage](#basic-usage)
- [Signed URLs](#signed-urls)
- [Srcset Generation](#srcset-generation)
- [Running Tests](#running-tests)
- [Features and bugs](#features-and-bugs)

## Install

```yaml
dependencies:
  imgix_client: ^1.0.0
```

## Basic Usage

A simple usage example:

```dart
import 'package:imgix_client/imgix_client.dart';

void main() {
  var urlBuilder = UrlBuilder('demos.imgix.net');
  var params = { 'w': '100', 'h': '100' };
  print('awesome: ${urlBuilder.createURL('bridge.png', params)}');
}

// Prints out:
// https://demos.imgix.net/bridge.png?h=100&w=100
```

HTTPS support is available by default. However, if you need HTTP support, call setUseHttps on the builder:

```dart
import 'package:imgix_client/imgix_client.dart';

void main() {
  var builder = UrlBuilder('demos.imgix.net');
  builder.useHttps = false; // use http

  var params = { 'w': '100', 'h': '100' };
  print(builder.createURL("bridge.png", params));
}


// Prints out
// http://demos.imgix.net/bridge.png?h=100&w=100
```

## Signed URLs

To produce a signed URL, you must enable secure URLs on your source and then
provide your signature key to the URL builder.

```dart
import 'package:imgix_client/imgix_client.dart';

void main() {
  var builder = UrlBuilder('demos.imgix.net');
  builder.useHttps = false; // use http
  builder.signKey = 'test1234'; // set sign key

  var params = { 'w': '100', 'h': '100' };
  print(builder.createURL("bridge.png", params));
}


// Prints out
// http://demos.imgix.net/bridge.png?h=100&w=100&s=bb8f3a2ab832e35997456823272103a4
```

## Srcset Generation

The imgix-java library allows for generation of custom `srcset` attributes, which can be invoked through `createSrcSet()`. By default, the `srcset` generated will allow for responsive size switching by building a list of image-width mappings.

```dart
import 'package:imgix_client/imgix_client.dart';

void main() {
  var ub = UrlBuilder('demos.imgix.net', true, 'my-token', false);
  var srcset = ub.createSrcSet("bridge.png");

  print(srcset);
}
```

Will produce the following attribute value, which can then be served to the client:

```html
https://demos.imgix.net/bridge.png?w=100&s=494158d968e94ac8e83772ada9a83ad1 100w,
https://demos.imgix.net/bridge.png?w=116&s=6a22236e189b6a9548b531330647ffa7 116w,
https://demos.imgix.net/bridge.png?w=134&s=cbf91f556dd67c0b9e26cb9784a83794 134w,
                                    ...
https://demos.imgix.net/bridge.png?w=7400&s=503e3ba04588f1c301863c9a5d84fe91 7400w,
https://demos.imgix.net/bridge.png?w=8192&s=152551ce4ec155f7a03f60f762a1ca33 8192w
```
In cases where enough information is provided about an image's dimensions, `createSrcSet()` will instead build a `srcset` that will allow for an image to be served at different resolutions. The parameters taken into consideration when determining if an image is fixed-width are `w` (width), `h` (height), and `ar` (aspect ratio). By invoking `createSrcSet()` with either a width **or** the height and aspect ratio (along with `fit=crop`, typically) provided, a different `srcset` will be generated for a fixed-size image instead.

```dart
import 'package:imgix_client/imgix_client.dart';

void main() {
  var ub = UrlBuilder('demos.imgix.net', true, 'my-token', false);
  var params = { 'h': '200', 'ar': '3:2', 'fit': 'crop' };
  var srcset = ub.createSrcSet("bridge.png", params);

  print(srcset);
}
```

Will produce the following attribute value:

```html
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=1&fit=crop&h=200&s=4c79373f535df7e2594a8f6622ec6631 1x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=2&fit=crop&h=200&s=dc818ae4522494f2f750651304a4d825 2x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=3&fit=crop&h=200&s=ba1ec0cef6c77ff02330d40cc4dae932 3x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=4&fit=crop&h=200&s=b51e497d9461be62354c0ea12b6524fb 4x,
https://demos.imgix.net/bridge.png?ar=3%3A2&dpr=5&fit=crop&h=200&s=dc37c1fbee505d425ca8e3764b37f791 5x
```

For more information to better understand `srcset`, we recommend [Eric Portis' "Srcset and sizes" article](https://ericportis.com/posts/2014/srcset-sizes/) which goes into depth about the subject.


## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/amondnet/imgix.dart/issues


## Running Tests

To run tests clone this project and run:

```
pub run test
```


CHANGELOG.md
## 3.0.0

- fix: an error is raised if params is immutable 
- feat: dart 3.0.0
- ci: use dart action
- build: updated the SDK lower-bound to 3.0.

## 2.0.0

- feat: null safety

## 1.0.0+3

- dependency update

## 1.0.0+2

- docs: example

## 1.0.0+1

- docs: basic usage

## 1.0.0

- Initial version


test/all_test.dart
import 'dart:collection';

import 'package:imgix_client/imgix_client.dart';
import 'package:test/test.dart';

void main() {
  group('srcSet', () {
    test('testURLBuilderRaisesExceptionOnNoDomains', () {
      expect(() => UrlBuilder(''), throwsArgumentError);
    });

    test('testURLBuilderUsesHttpsByDefault', () {
      // Test `URLBuilder` uses https by default.
      // This test uses the single-parameter constructor.
      // The single parameter constructor is the only constructor
      // where "default" makes sense. E.g. calling
      // `URLBuilder("example.imgix.net", true)`
      // passes `true` to the constructor explicitly.
      var ub = UrlBuilder('example.imgix.net');
      var expected = 'https://example.imgix.net/image/file.png?ixlib=dart-' +
          UrlBuilder.VERSION;
      expect(ub.createURL('image/file.png'), expected);
    });
    test('testSetUseHttpsFalse', () {
      // Test `setUseHttps` to `false`.
      var ub = UrlBuilder('example.imgix.net');
      ub.useHttps = false;
      var expected = 'http://example.imgix.net/image/file.png?ixlib=dart-' +
          UrlBuilder.VERSION;
      expect(ub.createURL('image/file.png'), expected);
    });
    test('testSetUseHttpsTrue', () {
      // Test `setUseHttps` to `false`.
      var ub = UrlBuilder('example.imgix.net');
      ub.useHttps = true;
      var expected = 'https://example.imgix.net/image/file.png?ixlib=dart-' +
          UrlBuilder.VERSION;
      expect(ub.createURL('image/file.png'), expected);
    });

    test('testHelperBuildAbsolutePath', () {
      var uh = UrlHelper(
          'securejackangers.imgix.net', "/example/chester.png", 'http');
      expect(
          'http://securejackangers.imgix.net/example/chester.png', uh.getURL());
    });

    test('testHelperBuildRelativePath', () {
      var uh = UrlHelper(
          'securejackangers.imgix.net', 'example/chester.png', 'http');
      expect(
          'http://securejackangers.imgix.net/example/chester.png', uh.getURL());
    });

    test('testHelperBuildNestedPath', () {
      var uh = new UrlHelper('securejackangers.imgix.net',
          'http://www.somedomain.com/example/chester.png', 'http');
      expect(
          'http://securejackangers.imgix.net/http%3A%2F%2Fwww.somedomain.com%2Fexample%2Fchester.png',
          uh.getURL());
    });

    test('testHelperBuildAbsolutePathWithParams', () {
      var uh = UrlHelper(
          'securejackangers.imgix.net', "/example/chester.png", 'http');
      uh.setParameter('w', 500);
      expect('http://securejackangers.imgix.net/example/chester.png?w=500',
          uh.getURL());
    });
    test('testHelperBuildRelativePathWithParams', () {
      var uh = UrlHelper(
          'securejackangers.imgix.net', "example/chester.png", 'http');
      uh.setParameter('w', 500);
      expect('http://securejackangers.imgix.net/example/chester.png?w=500',
          uh.getURL());
    });
    test('testHelperBuildNestedPathWithParams', () {
      var uh = new UrlHelper('securejackangers.imgix.net',
          'http://www.somedomain.com/example/chester.png', 'http');
      uh.setParameter('w', 500);

      expect(
        'http://securejackangers.imgix.net/http%3A%2F%2Fwww.somedomain.com%2Fexample%2Fchester.png?w=500',
        uh.getURL(),
      );
    });
    test('testHelperBuildSignedURLWithHashMapParams', () {
      var params = HashMap<String, String>();
      params['w'] = '500';

      var uh = UrlHelper('securejackangers.imgix.net', 'example/chester.png',
          'http', "Q61NvXIy", params);
      expect(
          'http://securejackangers.imgix.net/example/chester.png?w=500&s=787b9057d5c077fe168b4849737d8a90',
          uh.getURL());
    });
    test('testHelperBuildSignedURLWithHashSetterParams', () {
      var uh = UrlHelper('securejackangers.imgix.net', 'example/chester.png',
          'http', "Q61NvXIy");
      uh.setParameter("w", 500);

      expect(
          'http://securejackangers.imgix.net/example/chester.png?w=500&s=787b9057d5c077fe168b4849737d8a90',
          uh.getURL());
    });
    test('testHelperBuildSignedURLWithWebProxyWithNoEncoding', () {
      var uh = UrlHelper(
          'jackttl2.imgix.net',
          'http%3A%2F%2Fa.abcnews.com%2Fassets%2Fimages%2Fnavigation%2Fabc-logo.png%3Fr%3D20',
          'http',
          'JHrM2ezd');
      expect(
        'http://jackttl2.imgix.net/http%3A%2F%2Fa.abcnews.com%2Fassets%2Fimages%2Fnavigation%2Fabc-logo.png%3Fr%3D20?s=cf82defe3436a957262d0e64c21e72f9',
        uh.getURL(),
      );
    });
    test('testBuildSignedURLWithWebProxyWithUnencodedInput', () {
      var uh = UrlHelper(
          'imgix-library-web-proxy-test-source.imgix.net',
          'https://paulstraw.imgix.net/colon:test/benice.jpg',
          'https',
          'qN5VOqaLGQUFzETO');
      expect(
        'https://imgix-library-web-proxy-test-source.imgix.net/https%3A%2F%2Fpaulstraw.imgix.net%2Fcolon%3Atest%2Fbenice.jpg?s=175a054524d75840735855b9263be591',
        uh.getURL(),
      );
    });

    test('testBuilderWithFullyQualifiedURL', () {
      var ub =
          UrlBuilder('my-social-network.imgix.net', true, 'FOO123bar', false);
      expect(
        'https://my-social-network.imgix.net/http%3A%2F%2Favatars.com%2Fjohn-smith.png?s=493a52f008c91416351f8b33d4883135',
        ub.createURL('http://avatars.com/john-smith.png'),
      );
    });

    test('testBuilderWithFullyQualifiedURLAndParameters', () {
      var ub =
          UrlBuilder('my-social-network.imgix.net', true, 'FOO123bar', false);
      Map<String, String> params = HashMap<String, String>();
      params['w'] = '400';
      params['h'] = '300';
      expect(
        'https://my-social-network.imgix.net/http%3A%2F%2Favatars.com%2Fjohn-smith.png?h=300&w=400&s=a201fe1a3caef4944dcb40f6ce99e746',
        ub.createURL('http://avatars.com/john-smith.png', params),
      );
    });

    test('testHelperBuildSignedUrlWithIxlibParam', () {
      var domains = ['assets.imgix.net'];
      var ub = UrlBuilder('assets.imgix.net', true, '', true);
      expect(hasURLParameter(ub.createURL('/users/1.png'), 'ixlib'), isTrue);

      ub = UrlBuilder('assets.imgix.net', true, '', false);
      expect(hasURLParameter(ub.createURL('/users/1.png'), 'ixlib'), isFalse);
    });

    test('testTargetWidths', () {
      var actual = UrlBuilder.targetWidths(100, 8192, 8);
      var targetWidths = [
        100,
        116,
        135,
        156,
        181,
        210,
        244,
        283,
        328,
        380,
        441,
        512,
        594,
        689,
        799,
        927,
        1075,
        1247,
        1446,
        1678,
        1946,
        2257,
        2619,
        3038,
        3524,
        4087,
        4741,
        5500,
        6380,
        7401,
        8192
      ];
      for (var i = 0; i < targetWidths.length; ++i) {
        expect(actual[i], targetWidths[i]);
      }
    });

    test('testParamKeysAreEscaped', () {
      var params = HashMap<String, String>();
      params['hello world'] = 'interesting';

      var uh = UrlHelper('demo.imgix.net', 'demo.png', 'https', null, params);

      expect(uh.getURL(),
          'https://demo.imgix.net/demo.png?hello%20world=interesting');
    });

    test('testParamValuesAreEscaped', () {
      var params = HashMap<String, String>();
      params['hello_world'] = '/foo\"> <script>alert(\"hacked\")</script><';

      var uh = UrlHelper('demo.imgix.net', 'demo.png', 'https', null, params);

      expect(uh.getURL(),
          'https://demo.imgix.net/demo.png?hello_world=%2Ffoo%22%3E%20%3Cscript%3Ealert(%22hacked%22)%3C%2Fscript%3E%3C');
    });

    test('testBase64ParamVariantsAreBase64Encoded', () {
      var params = HashMap<String, String>();
      params['txt64'] = 'I cannøt belîév∑ it wors! 😱';

      var uh = UrlHelper('demo.imgix.net', '~text', 'https', null, params);

      expect(uh.getURL(),
          'https://demo.imgix.net/~text?txt64=SSBjYW5uw7h0IGJlbMOuw6l24oiRIGl0IHdvcu-jv3MhIPCfmLE');
    });

    test('testExtractDomain', () {
      var url = 'http://jackangers.imgix.net/chester.png';
      expect('jackangers.imgix.net', extractDomain(url));
    });

    test('testEncodeDecode', () {
      final url =
          'http://a.abcnews.com/assets/images/navigation/abc-logo.png?r=20';
      final encodedUrl =
          'http%3A%2F%2Fa.abcnews.com%2Fassets%2Fimages%2Fnavigation%2Fabc-logo.png%3Fr%3D20';

      expect(
        encodedUrl,
        UrlHelper.encodeURIComponent(url),
      );
      expect(url, UrlHelper.decodeURIComponent(encodedUrl));
      expect(
          encodedUrl,
          UrlHelper.encodeURIComponent(
              UrlHelper.decodeURIComponent(encodedUrl)));
    });

    test('testInvalidDomainAppendSlash', () {
      expect(() => UrlBuilder('test.imgix.net/'), throwsArgumentError);
    });

    test('testInvalidDomainPrependScheme', () {
      expect(() => UrlBuilder('https://test.imgix.net/'), throwsArgumentError);
    });

    test('testInvalidDomainAppendDash', () {
      expect(() => UrlBuilder('test.imgix.net-'), throwsArgumentError);
    });
  });
}

String extractDomain(String url) {
  try {
    var parsed = Uri.parse(url);
    var curDomain = parsed.authority;
    return curDomain;
  } catch (e) {
    print(e);
  }

  return '';
}

bool hasURLParameter(String url, String param) {
  try {
    var parsed = Uri.parse(url);
    var query = parsed.query;
    return query != null && query.contains(param);
  } catch (e) {
    return false;
  }
}


lib/src/url_helper.dart
import 'dart:collection';
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:quiver/strings.dart';

class UrlHelper {
  final String _domain;
  String _path;
  final String _scheme;
  final String? _signKey;
  final Map<String, String> _parameters;

  UrlHelper(this._domain, this._path,
      [this._scheme = 'http',
      this._signKey = '',
      Map<String, String> parameters = const {}])
      : _parameters = SplayTreeMap.of(parameters) {
    if (!_path.startsWith('/')) {
      _path = '/$_path';
    }
  }

  void setParameter(String key, dynamic value) {
    if (value is String || value is num) {
      var stringValue = value.toString();
      if (isNotBlank(stringValue)) {
        _parameters[key] = stringValue;
      } else {
        _parameters.remove(key);
      }
    }
  }

  void deleteParameter(String key) {
    setParameter(key, '');
  }

  // URL Safe base64
  String _encodeBase64(String str) {
    String b64EncodedString;

    try {
      var stringBytes = utf8.encode(str);
      b64EncodedString = base64UrlEncode(stringBytes);
      b64EncodedString = b64EncodedString.replaceAll('=', '');
    } catch (e) {
      throw ArgumentError(e);
    }

    return b64EncodedString;
  }

  String getURL() {
    var queryPairs = <String>[];

    for (var entry in _parameters.entries) {
      var k = Uri.encodeComponent(entry.key);
      var v = entry.value;

      String encodedValue;

      if (k.endsWith('64')) {
        encodedValue = _encodeBase64(v);
      } else {
        encodedValue = Uri.encodeComponent(v);
      }
      queryPairs.add('$k=$encodedValue');
    }

    var query = queryPairs.join('&');

    var decodedPath = Uri.decodeComponent(_path.substring(1));
    if (decodedPath.startsWith('http://') ||
        decodedPath.startsWith('https://')) {
      _path = '/${Uri.encodeComponent(decodedPath)}';
    }

    if (isNotBlank(_signKey)) {
      var delim = query == '' ? '' : '?';
      var toSign = _signKey! + _path + delim + query;
      var signature = md5.convert(utf8.encode(toSign)).toString();

      if (query.isNotEmpty) {
        query += '&s=$signature';
      } else {
        query = 's=$signature';
      }

      return buildURL(_scheme, _domain, _path, query);
    }

    return buildURL(_scheme, _domain, _path, query);
  }

  ///////////// Static

  static String buildURL(
      String scheme, String host, String path, String query) {
    // do not use URI to build URL since it will do auto-encoding which can break our previous signing
    var url = '$scheme://$host$path?$query';
    if (url.endsWith('#')) {
      url = url.substring(0, url.length - 1);
    }

    if (url.endsWith('?')) {
      url = url.substring(0, url.length - 1);
    }

    return url;
  }

  static String convertMD5(String input) {
    var array = utf8.encode(input);
    return md5.convert(array).toString();
  }

  static String encodeURIComponent(String s) {
    String result;

    try {
      result = Uri.encodeComponent(s)
          .replaceAll('\\+', '%20')
          .replaceAll('\\%21', '!')
          .replaceAll('\\%27', "'")
          .replaceAll('\\%28', '(')
          .replaceAll('\\%29', ')')
          .replaceAll('\\%7E', '~');
    } catch (e) {
      result = s;
    }

    return result;
  }

  static String decodeURIComponent(String s) {
    String result;

    try {
      result = Uri.decodeComponent(s);
    } catch (e) {
      result = s;
    }

    return result.toString();
  }
}


lib/src/url_builder.dart
import 'dart:collection';

import 'package:built_collection/built_collection.dart';
import 'package:quiver/strings.dart';

import 'url_helper.dart';

class UrlBuilder {
  static final String version = '1.0.0';
  static final String domainRegex =
      '^(?:[a-z\\d\\-_]{1,62}\\.){0,125}(?:[a-z\\d](?:\\-(?=\\-*[a-z\\d])|[a-z]|\\d){0,62}\\.)[a-z\\d]{1,63}\$';

  static final srcsetTargetWidths = [
    100,
    116,
    135,
    156,
    181,
    210,
    244,
    283,
    328,
    380,
    441,
    512,
    594,
    689,
    799,
    927,
    1075,
    1247,
    1446,
    1678,
    1946,
    2257,
    2619,
    3038,
    3524,
    4087,
    4741,
    5500,
    6380,
    7401,
    8192
  ];
  static final List<int> sss = <int>[];
  static const int srcsetWidthTolerance = 8;
  static const int minWidth = 100;
  static const int maxWidth = 8192;
  static final dprQualities = [75, 50, 35, 23, 20];
  static final targetRatios = [1, 2, 3, 4, 5];

  final String domain;
  final bool useHttps;
  String? signKey;
  final bool includeLibraryParam;

  UrlBuilder(this.domain,
      {this.useHttps = true, this.signKey, this.includeLibraryParam = true}) {
    final domainPattern = RegExp(domainRegex);
    if (isBlank(domain)) {
      throw ArgumentError.value(
          domain, 'domain', 'At lease one domain must be passed to URLBuilder');
    } else if (!domainPattern.hasMatch(domain)) {
      throw ArgumentError.value(domain, 'domain',
          'Domain must be passed in as a fully-qualified domain name and should not include a protocol or any path element, i.e. "example.imgix.net".');
    }
  }

  String createURL(String path, [Map<String, String>? params]) {
    final newParam = Map.of(params ??= {});
    final scheme = useHttps ? 'https' : 'http';
    if (includeLibraryParam) {
      newParam['ixlib'] = 'dart-$version';
    }
    return UrlHelper(domain, path, scheme, signKey, newParam).getURL();
  }

  static BuiltList<int> targetWidths(int begin, int end, int tol) {
    return computeTargetWidths(
        begin.toDouble(), end.toDouble(), tol.toDouble());
  }

  /// Create a srcset given a `path` and a map of `params`.
  ///
  /// This function creates a dpr based srcset if `params`
  /// contain either:
  /// - a width "w" param, _or_
  /// - a height "h" and aspect ratio "ar" params
  ///
  /// Otherwise, a srcset of width-pairs is created.
  /// path - path to the image, i.e. "image/file.png"
  /// @param params - map of query parameters
  /// @param tol - tolerable amount of width value variation
  /// @param disableVariableQuality - flag to toggle variable image
  //  output quality.
  ///
  /// @return srcset attribute string
  ///

  String createSrcSet(String path,
      {Map<String, String>? params,
      int begin = minWidth,
      int end = maxWidth,
      int tol = srcsetWidthTolerance,
      bool disableVariableQuality = false,
      Iterable<int>? targets}) {
    final srcsetParams = SplayTreeMap<String, String>.of(params ?? {});
    if (targets?.isNotEmpty == true) {
      return createSrcSetPairs(path, srcsetParams, targets!);
    } else if (isDpr(srcsetParams)) {
      return createSrcSetDPR(path, srcsetParams, disableVariableQuality);
    } else {
      var targets = targetWidths(begin, end, tol);
      return createSrcSetPairs(path, srcsetParams, targets);
    }
  }

  String createSrcSetDPR(
      String path, Map<String, String> params, bool disableVariableQuality) {
    final srcset = StringBuffer();
    final srcsetParams = HashMap<String, String>.of(params);

    final hasQuality = srcsetParams['q'] != null;

    for (final ratio in targetRatios) {
      srcsetParams['dpr'] = ratio.toString();

      if (!disableVariableQuality && !hasQuality) {
        srcsetParams['q'] = dprQualities[ratio - 1].toString();
      }
      srcset.write(createURL(path, srcsetParams));
      srcset.write(' ');
      srcset.write(ratio);
      srcset.write('x,\n');
    }
    var string = srcset.toString();
    return string.substring(0, string.length - 2);
  }

  /// Create an `ArrayList` of integer target widths.
  ///
  /// This function is the implementation details of `targetWidths`.
  /// This function exists to provide a consistent interface for
  /// callers of `targetWidths`.
  ///
  /// This function implements the syntax that fulfills the semantics
  /// of `targetWidths`. Meaning, `begin`, `end`, and `tol` are
  /// to be whole integers, but computation requires `double`s. This
  /// function hides this detail from callers.
  static BuiltList<int> computeTargetWidths(
      double begin, double end, double tol) {
    if (_notCustom(begin, end, tol)) {
      return BuiltList<int>.of(srcsetTargetWidths);
    }

    final resolutions = <int>[];
    if (begin == end) {
      // `begin` has not been mutated; cast back to `int`.
      resolutions.add(begin.toInt());
      return BuiltList.of(resolutions);
    }

    while (begin < end && begin < maxWidth) {
      // Round values so that the resulting `int` is truer
      // to expectations (i.e. 115.99999 --> 116).
      resolutions.add(begin.round());
      begin *= 1 + (tol / 100) * 2;
    }

    var lastIndex = resolutions.length - 1;
    if (resolutions[lastIndex] < end) {
      // `end` has not been mutated; cast back to `int`.
      resolutions.add(end.toInt());
    }

    return BuiltList.of(resolutions);
  }

  bool isDpr(Map<String, String> params) {
    var width = params['w'];
    var hasWidth = width != null && isNotBlank(params['w']);

    var height = params['h'];
    var hasHeight = (height != null) && height.isNotEmpty;

    var aspectRatio = params['ar'];
    var hasAspectRatio = (aspectRatio != null) && aspectRatio.isNotEmpty;

    // If `params` have a width param or _both_ height and aspect
    // ratio parameters then the srcset to be constructed with
    // these params _is dpr based_.
    return hasWidth || (hasHeight && hasAspectRatio);
  }

  String createSrcSetPairs(
      String path, Map<String, String> params, Iterable<int> targets) {
    final srcset = StringBuffer();

    final srcSetParams = Map.of(params);

    for (final width in targets) {
      srcSetParams['w'] = width.toString();
      srcset.write(createURL(path, srcSetParams));
      srcset.write(' ');
      srcset.write(width);
      srcset.writeln('w,');
    }

    var string = srcset.toString();
    return string.substring(0, srcset.length - 2);
  }

  static bool _notCustom(double begin, double end, double tol) {
    var defaultBegin = (begin == minWidth);
    var defaultEnd = (end == maxWidth);
    var defaultTol = (tol == srcsetWidthTolerance);

// A list of target widths is _NOT_ custom if `begin`, `end`,
// and `tol` are equal to their default values.
    return defaultBegin && defaultEnd && defaultTol;
  }
}
